python/ray/tune/analysis/post_experiment.py

```diff
@@ -0,0 +1,156 @@
+        return cfg_dict
+    except Exception:
+        return None
+
+def load_results_to_df(directory, result_name="result.json"):
+    """ Loads results to pandas dataframe """
+    exp_directories = [
+        dirpath for dirpath, dirs, files in os.walk(directory) for f in files
+        if f == result_name
+    ]
+    data = [_resolve(d, result_name) for d in exp_directories]
+    data = [d for d in data if d]
+    return pd.DataFrame(data)
+
+def generate_plotly_dim_dict(df, field):
+    dim_dict = {}
+    dim_dict["label"] = field
+    column = df[field]
+    if is_numeric_dtype(column):
+        dim_dict["values"] = column
+    elif is_string_dtype(column):
+        texts = column.unique()
+        dim_dict["values"] = [
+            np.argwhere(texts == x).flatten()[0] for x in column
+        ]
+        dim_dict["tickvals"] = list(range(len(texts)))
+        dim_dict["ticktext"] = texts
+    else:
+        raise Exception("Unidentifiable Type")
+
+    return dim_dict
+
+
+def get_result_path(trial_dir):
+    return os.path.join(trial_dir, "progress.csv")
+
+
+def get_result_backup_path(result_path):
+    return f"{result_path}.backup"
+
+
+def result_backed_up(trial_dir):
+    result_path = get_result_path(trial_dir=trial_dir)
+    result_backup_path = get_result_backup_path(result_path=result_path)
+
+    return os.path.exists(result_backup_path)
+
+
+def backup_result(trial_dir):
+    result_path = get_result_path(trial_dir=trial_dir)
+    result_backup_path = get_result_backup_path(result_path=result_path)
+    copyfile(result_path, result_backup_path)
+
+
+def prune_restore_dataframes(dataframes):
+    if len(dataframes) < 2:
+        return dataframes[0]
+
+    dataframe1 = dataframes[0]
+    dataframe2 = prune_restore_dataframes(dataframes[1:])
+
+    return max(dataframe1, dataframe2, key=lambda d: d.shape[0])
+
+
+def clean_dataframe(dataframe):
+    restore_begins = tuple(
+        np.where(dataframe['iterations_since_restore'] == 1)[0])
+    restore_ends = (*restore_begins[1:], dataframe.shape[0])
+    restore_dataframes_index = tuple(zip(restore_begins, restore_ends))
+    restore_dataframes = tuple(
+        dataframe[slice(*restore_dataframe_index)]
+        for restore_dataframe_index in restore_dataframes_index)
+
+    cleaned_dataframe = prune_restore_dataframes(restore_dataframes)
+
+    return cleaned_dataframe
+
+
+def clean_trial(trial_dir):
+    result_path = get_result_path(trial_dir=trial_dir)
+    result_backup_path = get_result_backup_path(result_path=result_path)
+
+    print(result_backup_path)
+    dataframe = pd.read_csv(result_backup_path)
+
+    cleaned_dataframe = clean_dataframe(dataframe)
+
+    cleaned_dataframe.to_csv(result_path, index=False)
+
+
+def fix_ray_results(experiment_dir):
+    trial_dirs = [
+        os.path.join(experiment_dir, trial_dir)
+        for trial_dir in next(os.walk(experiment_dir))[1]
+    ]
+
+    for trial_dir in trial_dirs:
+        if not result_backed_up(trial_dir):
```

Comment:
It looks like these functions are the same as in my gist. I remember that there's some bug here related to the backing up, but can't remember exactly what it was. Can you verify that the following case produces expected result:
1. Clean experiment that is in progress.
2. Let the experiment run to the end.
3. Clean experiment again.

Reply:
Also, adding a test case for that would be great.

Reply:
Hi @hartikainen, thank you for the review. When testing, would experiment_dir be something like "path/to/ray_results/<experiment_name>"?

Reply:
Yep, `path/to/ray_results/<experiment_name>` sounds correct.

Reply:
After playing around with running/cleaning experiments (including the case you outlined) it seems that the only error I encountered was pandas raising an error when trying to open empty progress.csv files. I changed the function so that we return in that case b/c there is nothing to clean.

Not sure why experiments have trials with empty progress.csv files (from my testing, anywhere from 0-20 trials per experiment). Is this correct behavior?

Reply:
> After playing around with running/cleaning (including the case you outlined) it seems that the only error I encountered was pandas raising an error when trying to open empty progress.csv files. I changed the function so that we return in that case b/c there is nothing to clean.

This sounds like the correct thing to do.

> Not sure why some experiments have empty progress.csv files (from my testing, anywhere from 0-20 experiments per tune.run_experiments example). Is this correct behavior?

 Assuming that those trials have actually taken > 0 steps, then this does not sound correct. However, maybe it's possible that if you launch an experiment and it crashes due to some error, then it's `progress.csv` ends up being empty? Can you verify this?

Reply:
We ended up taking out this functionality from this PR.